src/resolve/target.rs

```diff
@@ -0,0 +1,331 @@
+use crate::diagnostics::Diagnostics;
+use crate::error::{Error, Result};
+use crate::frontend::range::SourceRange;
+use crate::resolve::remote::{ContentType, Error as RemoteError, Remote};
+use crate::resolve::{Command, Context, Include, Permissions, ContextType};
+
+/// Target pointed to by URL before the permission check.
+#[must_use]
+#[derive(Debug)]
+pub struct Target<'a, 'd> {
+    inner: TargetInner,
+    meta: Meta<'a, 'd>,
+}
+
+/// Target after canonicalization
+#[must_use]
+#[derive(Debug)]
+pub struct TargetCanonicalized<'a, 'd> {
+    inner: TargetInner,
+    meta: Meta<'a, 'd>,
+}
+
+/// Target after its permissions have been checked
+#[must_use]
+#[derive(Debug)]
+pub struct TargetChecked<'a, 'd> {
+    inner: TargetInner,
+    meta: Meta<'a, 'd>,
+}
+
+#[derive(Debug)]
+struct Meta<'a, 'd> {
+    url: Url,
+    context: &'a Context,
+    project_root: &'a Path,
+    permissions: &'a Permissions,
+    range: SourceRange,
+    diagnostics: &'a Diagnostics<'d>,
+}
+
+#[derive(Debug)]
+enum TargetInner {
+    /// Implemented commands / codegen.
+    ///
+    /// Ex: `![](//TOC)`
+    Implementation(String),
+    /// Local file inside the project root or the context directory.
+    ///
+    /// The `PathBuf` must be relative.
+    ///
+    /// Ex: `![](/foo.md)`, `![](foo.md)`
+    LocalRelative(PathBuf),
+    /// Any file with an absolute path.
+    ///
+    /// Ex: `![](file:///foo.md)`
+    LocalAbsolute(PathBuf),
+    /// Remote source / file.
+    ///
+    /// Ex: `![](https://foo.bar/baz.md)`
+    Remote(Url),
+}
+
+impl<'a, 'd> Target<'a, 'd> {
+    /// Create a new Target for the given URL, resolved in the given context.
+    ///
+    /// This target can be canonicalized and access-checked within the context before being converted
+    /// to the respective Include.
+    /// Local relative files are resolved relative to the project_root.
+    pub fn new(
+        url: &str, context: &'a Context, project_root: &'a Path, permissions: &'a Permissions,
+        range: SourceRange, diagnostics: &'a Diagnostics<'d>,
+    ) -> Result<Target<'a, 'd>> {
+        let url = match context.url.join(url) {
+            Ok(url) => url,
+            Err(err) => {
+                diagnostics
+                    .error("couldn't resolve file")
+                    .with_error_section(range, "defined here")
+                    .note(format!("tried to resolve {}", url))
+                    .note(format!("malformed reference: {}", err))
+                    .emit();
+                return Err(Error::Diagnostic);
+            },
+        };
+        let inner = match url.scheme() {
+            "heradoc" => match url.domain() {
+                Some("document") => TargetInner::LocalRelative(url.path_segments().unwrap().collect()),
+                Some(domain) => TargetInner::Implementation(domain.to_string()),
+                None => {
+                    diagnostics
+                        .error("no heradoc implementation domain found")
```

Comment:
That seems unclear to me. We didn't find the specified one, not none at all.
```suggestion
                        .error("heradoc implementation domain not found")
```